Review Code, Add Tests and BinghamMetric Class instance, and first version of tutorial.

RafaelNH opened the pull request:
Hi @CHrlS98,

Here is my input for the Bingham fitting functions. 

What I've done:

1. I reviewed all functions and improved their documentation

2. As I promised, I added the test
3. As I also promised, I added the BinghamMetric Class instance in the function to the fit Binghams in multiple voxels - to be more efficient in dealing with different Bingham metrics. 
4. Created a first version of the tutorial with the description of some of the Bingham metrics.

Two notes:

- We may consider increasing the default value for max_search_angle. the tests were not passing with the current default value. Indeed, you can notice that in the test I am using a larger value for max_search_angle (45 degrees)
- (this may be a comment for Julio) Regarding the ´bingham_from_sh´ function, it is not recommended to load data and processes computation in the same function. Therefore, I've developed a new function named ´bingham_from_odf´ to perform Bingham fitting directly from ODF maps. I understand there might be concerns regarding the memory demands of handling ODF maps directly. To address this, we could consider offering an alternative function that operates on spherical harmonics (that should be inputted not loaded). If you believe incorporating spherical harmonics input would be beneficial, please feel free to adjust ´bingham_from_odf´ accordingly and update the example.

villalonreina commented:
@RafaelNH Thanks so much for all this work. Yes, I agree that it's better to fit Bingham directly on the ODF maps. It makes sense to have the `bingham_from_odf` function instead of `bingham_from_sh`. Maybe we can have one example/tutorial where a spherical harmonics file is inputted, then converted to odf (with `sh_to_sf`), and then call `bingham_from_odf`. What do you think? The thing is that I think that in practice most researchers will fit their ODFs for other purposes (tractography most likely) and will save the spherical harmonics to disk (likely in Nifti files), which then could be reused to fit the Bingham metrics.

RafaelNH commented:
@CHrlS98 As you requested I rebased the code according to your latest branch version. 

@villalonreina I added now the function to fit "Binghams" from spherical harmonics and include some notes in documentation.

I think I am done here. Feel free to review the code and merge it to DIPY's PR if you don't detect any issues. After that we can notify the DIPY team that this is ready to be merged and ask for their input.

CHrlS98 commented on dipy/direction/bingham.py:
```diff
@@ -115,134 +55,279 @@ def _bingham_fit_peak(sf, peak, sphere, max_angle):
+    where $f(n)$ is the Bingham function value at a given direction $n$, $f_0%
+    is the Bingham maximum peak value, $k_1$ and $k_2$ are the concentration
+    constants parameters (large values correspond to lower dispersion values)
+    along the two dispersion axes $\mu_1$ and $\mu_2$-
```
```suggestion
    where $f(n)$ is the Bingham function value at a given direction $n$, $f_0$
    is the Bingham maximum peak value, $k_1$ and $k_2$ are the concentration
    constants parameters (large values correspond to lower dispersion values)
    along the two dispersion axes $\mu_1$ and $\mu_2$.
```

CHrlS98 commented on dipy/direction/bingham.py:
```diff
@@ -254,218 +339,434 @@ def bingham_fiber_density(bingham_fits, n_thetas=50, n_phis=100):
-    and loops through the image to fit the Bingham distribution at each voxel.
+def odi2k(odi):
+    """ Convent Bingham/Watson concentration parameter k to the orientation
+    dispersion index (ODI)
```
```suggestion
    """ Convert Bingham/Watson concentration parameter k to the orientation
    dispersion index (ODI)
```

CHrlS98 commented on dipy/direction/bingham.py:
```diff
@@ -115,134 +55,279 @@ def _bingham_fit_peak(sf, peak, sphere, max_angle):
+            concentration parameters k1 and k2 (indexes 1 and 2);
+            elements of Bingham main direction (indexes 3-5);
+            elements of Bingham dispersion major axis (indexes 6-8);
+            and elements of Bingham dispersion minor axis (indexs 9-11).
```
```suggestion
            and elements of Bingham dispersion minor axis (indexes 9-11).
```

CHrlS98 commented on dipy/direction/bingham.py:
```diff
@@ -115,134 +55,279 @@ def _bingham_fit_peak(sf, peak, sphere, max_angle):
+            concentration parameters k1 and k2 (indexes 1 and 2);
+            elements of Bingham main direction (indexes 3-5);
+            elements of Bingham dispersion major axis (indexes 6-8);
+            and elements of Bingham dispersion minor axis (indexs 9-11).
```
```suggestion
            and elements of Bingham dispersion minor axis (indexes 9-11).
```

CHrlS98 commented on dipy/direction/bingham.py:
```diff
@@ -254,218 +339,434 @@ def bingham_fiber_density(bingham_fits, n_thetas=50, n_phis=100):
-    concentration parameters (k1, k2) as described in [2]_ and [3]_.
-    
+def k2odi(k):
+    """ Convent Bingham/Watson concentration parameter k to the orientation
```
```suggestion
    """ Convert Bingham/Watson concentration parameter k to the orientation
```

CHrlS98 commented on dipy/direction/bingham.py:
```diff
@@ -254,218 +339,434 @@ def bingham_fiber_density(bingham_fits, n_thetas=50, n_phis=100):
+       tuple of nl elements containting the Bingham function parameters
+       in the following order:
+            Maximum value of the Bingham function (f0);
+            concentration parameters (k1 and k2;
```
```suggestion
            concentration parameters (k1 and k2);
```

CHrlS98 commented on dipy/direction/bingham.py:
```diff
@@ -254,218 +339,434 @@ def bingham_fiber_density(bingham_fits, n_thetas=50, n_phis=100):
+            concentration parameters k1 and k2 (indexes 1 and 2);
+            elements of Bingham main direction (indexes 3-5);
+            elements of Bingham dispersion major axis (indexes 6-8);
+            and elements of Bingham dispersion minor axis (indexs 9-11).
```
```suggestion
            and elements of Bingham dispersion minor axis (indexes 9-11).
```

CHrlS98 commented on dipy/direction/bingham.py:
```diff
@@ -254,218 +339,434 @@ def bingham_fiber_density(bingham_fits, n_thetas=50, n_phis=100):
+
+def global_voxel_metric(bmetric, bfd):
+    """ Compute a global scalar maps for metrics of Bingham functions
+    fitted to multiple ODF lobel
```
```suggestion
    fitted to multiple ODF lobes
```

CHrlS98 commented on dipy/direction/bingham.py:
```diff
@@ -254,218 +339,434 @@ def bingham_fiber_density(bingham_fits, n_thetas=50, n_phis=100):
+            concentration parameters k1 and k2 (indexes 1 and 2);
+            elements of Bingham main direction (indexes 3-5);
+            elements of Bingham dispersion major axis (indexes 6-8);
+            and elements of Bingham dispersion minor axis (indexs 9-11).
```
```suggestion
            and elements of Bingham dispersion minor axis (indexes 9-11).
```

CHrlS98 commented on dipy/direction/bingham.py:
```diff
@@ -254,218 +339,434 @@ def bingham_fiber_density(bingham_fits, n_thetas=50, n_phis=100):
+
+    @auto_attr
+    def kappa_total(self):
+        """Comnbined concentration parameters for each ODF lobe.
```
```suggestion
        """Combined concentration parameters for each ODF lobe.
```

CHrlS98 commented on dipy/direction/bingham.py:
```diff
@@ -254,218 +339,434 @@ def bingham_fiber_density(bingham_fits, n_thetas=50, n_phis=100):
+    @auto_attr
+    def odi_1(self):
+        """Orientation Dispersion index 1 computed for each ODF lobe from
+        concentration paramenter kappa_1."""
```
```suggestion
        concentration parameter kappa_1."""
```

CHrlS98 commented on dipy/direction/bingham.py:
```diff
@@ -254,218 +339,434 @@ def bingham_fiber_density(bingham_fits, n_thetas=50, n_phis=100):
+    @auto_attr
+    def odi_2(self):
+        """Orientation Dispersion index 1 computed for each ODF lobe from
+        concentration paramenter kappa_2."""
```
```suggestion
        concentration parameter kappa_2."""
```

CHrlS98 commented on doc/examples/reconst_bingham.py:
```diff
@@ -0,0 +1,304 @@
+# BinghamMetric can also be used to compute the averaged ODI quantities across
+# all ODF lobes (see below). The averaged quantitaties are computed by
+# weigthing each ODF lobe with their respective FD value. These quantities
+# are ploted in the following figure.
```
```suggestion
# are plotted in the following figure.
```

CHrlS98 reviewed (commented):
I read the comments and tutorial and noted some typos. I added them as suggestions. Otherwise it looks good to me, thanks a lot!

RafaelNH commented:
Thanks for spotting these typos. I've incorporated all of them

CHrlS98 commented:
Thank you, I'll go ahead with the merge.

villalonreina commented:
@CHrlS98 Is it merged already?

CHrlS98 commented:
@villalonreina yes, you can review the DIPY PR directly: https://github.com/dipy/dipy/pull/2826
